prov/gni/src/gnix_cm.c

```diff
@@ -53,122 +111,798 @@
+
+	COND_RELEASE(ep_priv->requires_lock, &ep_priv->vc_lock);
+
+	GNIX_DEBUG(FI_LOG_EP_CTRL, "Sent conn resp: %p\n", ep_priv);
+
+	return FI_SUCCESS;
+
+err_eq_write:
+err_write:
+err_smsg_init:
+	_gnix_mbox_free(ep_priv->vc->smsg_mbox);
+	ep_priv->vc->smsg_mbox = NULL;
+err_mbox_alloc:
+	_gnix_vc_destroy(ep_priv->vc);
+	ep_priv->vc = NULL;
+err_unlock:
+	COND_RELEASE(ep_priv->requires_lock, &ep_priv->vc_lock);
+
+	return ret;
 }
 
-DIRECT_FN STATIC int gnix_getpeer(struct fid_ep *ep, void *addr, size_t *addrlen)
+DIRECT_FN STATIC int gnix_shutdown(struct fid_ep *ep, uint64_t flags)
 {
 	return -FI_ENOSYS;
 }
 
-DIRECT_FN STATIC int gnix_connect(struct fid_ep *ep, const void *addr,
-			   const void *param, size_t paramlen)
+struct fi_ops_cm gnix_ep_msg_ops_cm = {
+	.size = sizeof(struct fi_ops_cm),
+	.setname = gnix_setname,
+	.getname = gnix_getname,
+	.getpeer = gnix_getpeer,
+	.connect = gnix_connect,
+	.listen = fi_no_listen,
+	.accept = gnix_accept,
+	.reject = fi_no_reject,
+	.shutdown = gnix_shutdown,
+};
+
+/******************************************************************************
+ *
+ * Passive endpoint handling
+ *
+ *****************************************************************************/
+
+/* Process an incoming connection request at a listening PEP. */
+static int __gnix_pep_connreq(struct gnix_fid_pep *pep, int fd)
 {
-	return -FI_ENOSYS;
+	int ret;
+	struct gnix_pep_sock_conn *conn;
+	struct fi_eq_cm_entry eq_entry;
+
+	/* Create and initialize a new connection request. */
+	conn = calloc(1, sizeof(*conn));
+	if (!conn) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Failed to alloc accepted socket conn\n");
+		return -FI_ENOMEM;
+	}
+
+	conn->fid.fclass = FI_CLASS_CONNREQ;
+	conn->fid.context = pep;
+	conn->sock_fd = fd;
+
+	/* Pull request data from the listening socket. */
+	conn->bytes_read += read(fd, &conn->req, sizeof(conn->req));
+	if (conn->bytes_read != sizeof(conn->req)) {
+		/* TODO Wait for more bytes. */
+		GNIX_FATAL(FI_LOG_EP_CTRL, "Unexpected read size\n");
+	}
+
+	conn->req.info.src_addr = &conn->req.src_addr;
+	conn->req.info.dest_addr = &conn->req.dest_addr;
+	conn->req.info.tx_attr = &conn->req.tx_attr;
+	conn->req.info.rx_attr = &conn->req.rx_attr;
+	conn->req.info.ep_attr = &conn->req.ep_attr;
+	conn->req.info.domain_attr = &conn->req.domain_attr;
+	conn->req.info.fabric_attr = &conn->req.fabric_attr;
+	conn->req.info.domain_attr->name = NULL;
+	conn->req.info.fabric_attr->name = NULL;
+	conn->req.info.fabric_attr->prov_name = NULL;
+
+	conn->info = &conn->req.info;
+	conn->info->handle = &conn->fid;
+
+	eq_entry.fid = &pep->pep_fid.fid;
+	eq_entry.info = fi_dupinfo(conn->info);
+
+	/* Tell user of a new conn req via the EQ. */
+	ret = fi_eq_write(&pep->eq->eq_fid, FI_CONNREQ, &eq_entry,
+			  sizeof(eq_entry), 0);
+	if (ret != sizeof(eq_entry)) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "fi_eq_write failed, err: %d\n", ret);
+		fi_freeinfo(conn->info);
+		free(conn);
+		return ret;
+	}
+
```

Comment:
The event queue event has a reference to the conn structure.  The user is supposed to use pieces of the EQE to create an EP and then call fi_accept on it.  The conn is cleaned up inside the fi_accept call.  The conn is not currently cleaned up if the user does not follow through accepting the connection.  I was saving that error handling/resiliency for a future commit.